library_samples/Python3/ocs_sample_library_preview/Authentication.py

```diff
@@ -0,0 +1,148 @@
+
+from .SdsError import SdsError
+
+import base64
+import configparser
+import hashlib
+from http.server import HTTPServer, BaseHTTPRequestHandler
+import json
+import os 
+import secrets
+import time
+from urllib.parse import urlparse, parse_qs
+import webbrowser
+
+import requests
+
+
+class Authentication(object):
+
+    def __init__(self, tenant, url, clientId, clientSecret):
+        self.__tenant = tenant
+        self.__clientId = clientId
+        self.__clientSecret = clientSecret
+        self.__url = url
+
+        self.__expiration = 0
+        self.__token = ""
+        if clientSecret is not None:
+            self.__getToken = self.__getClientIDSecretToken
+        else:
+            self.__getToken = self.__getPKCEToken
+        
+
+    def getToken(self):
+        return self.__getToken()
+
+
+    def __getClientIDSecretToken(self):
+           
+        if ((self.__expiration - time.time()) > 5 * 60):
+            return self.__token
+
+        tokenEndpoint = self.__url + "/identity/connect/token"
+
+        tokenInformation = requests.post(
+            tokenEndpoint,
+            data={"client_id": self.__clientId,
+                  "client_secret": self.__clientSecret,
+                  "grant_type": "client_credentials"})
+
+        token = json.loads(tokenInformation.content)
+
+        expiration = token.get("expires_in", None)
+        if expiration is None:
+            raise SdsError(
+                f"Failed to get token, check client id/secret: {token['error']}")
+
+        self.__expiration = float(expiration) + time.time()
+        self.__token = token['access_token']
+        return self.__token
+
+    def __getPKCEToken(self):
+
+        if ((self.__expiration - time.time()) > 5 * 60):
+            return self.__token
+
+        try:
+            debug = os.environ.get("OCS_PKCE_DEBUG", None)
```

Comment:
Could we make `debug` here a setting or a parameter passed into the sample? I'm not sure I'd use an environment variable for this.

Reply:
Since there is no other debug code anywhere I didn't want to change the interface of OCSClient to debug this single section of the code. It's your lib, your call :)

Reply:
i know i have seen people turn on what Urls we are sending to do some debugging and understanding what is happening .  I wonder if a debug parameter (or debug level) in the baseclient or somewhere in base of the lib makes sense and we can start to use it.  Obviously most of this  would be outside the scope of this PR and request, but might be worth thinking about

Reply:
From a notebook point of view it looks like this: 
![image](https://user-images.githubusercontent.com/25752437/80731526-0a166e80-8ad9-11ea-9aa2-4e3aa9ad6ac3.png)

Reply:
I think I would prefer leaving the step messages all the time, except maybe not displaying the token. So we can remove "debug". Deal?

Reply:
Seems like a good compromise, maybe? We can adjust later if needed, it's not really necessary to completely resolve that in this PR. I assume you'd still keep the last (Step 6) message, just don't log the whole token to the console.

Reply:
Yes, this is what I had in mind for step 6. My goal would be to have a quick 1_6 release so we can start using it.